src/main/haskell/kore/src/Kore/Step/OrOfExpandedPattern.hs

```diff
@@ -0,0 +1,383 @@
+traverseWithPairs mapper patt = do
+    mapped <- traverse mapper patt
+    return (filterOr (fmap fst mapped), extract (fmap snd mapped))
+
+{-| 'fmapFlattenWithPairs' fmaps an or in a similar way to
+'traverseFlattenWithPairs'.
+-}
+fmapFlattenWithPairs
+    ::  (  ExpandedPattern level variable
+        -> (OrOfExpandedPattern level variable, a)
+        )
+    -> OrOfExpandedPattern level variable
+    -> (OrOfExpandedPattern level variable, [a])
+fmapFlattenWithPairs mapper patt =
+    (flatten (fmap fst mapped), extract (fmap snd mapped))
+  where
+    mapped = fmap mapper patt
+
+{-| 'traverseFlattenWithPairs' is similar to 'traverseWithPairs', except
+that its function returns an 'OrOfExpandedPattern', so the actual result
+is flattened at the end.
+-}
+traverseFlattenWithPairs
+    :: Monad f
+    =>  (  ExpandedPattern level variable
+        -> f (OrOfExpandedPattern level variable, a)
+        )
+    -> OrOfExpandedPattern level variable
+    -> f (OrOfExpandedPattern level variable, [a])
+traverseFlattenWithPairs mapper patt = do
+    mapped <- traverse mapper patt
+    return (flatten (fmap fst mapped), extract (fmap snd mapped))
+
+{-| 'traverseFlattenWithPairsGeneric' is similar to 'traverseFlattenWithPairs',
+except that it works with 'MultiOr's.
+-}
+traverseFlattenWithPairsGeneric
+    :: Monad f
+    =>  (  a
+        -> f (OrOfExpandedPattern level variable, pair)
+        )
+    -> MultiOr a
+    -> f (OrOfExpandedPattern level variable, [pair])
+traverseFlattenWithPairsGeneric mapper patt = do
+    mapped <- traverse mapper patt
+    return (flatten (fmap fst mapped), extract (fmap snd mapped))
+
+{-| 'filterGeneric' simplifies a MultiOr according to a function which
+evaluates its children to true/false/unknown.
+-}
+filterGeneric
+    :: (child -> OrBool)
+    -> MultiOr child
+    -> MultiOr child
+filterGeneric orFilter (MultiOr patts) =
+    go orFilter [] patts
+  where
+    go  :: (child -> OrBool)
+        -> [child]
+        -> [child]
+        -> MultiOr child
+    go _ filtered [] = MultiOr (reverse filtered)
+    go filterOr' filtered (element:unfiltered) =
+        case filterOr' element of
+            OrTrue -> MultiOr [element]
+            OrFalse -> go filterOr' filtered unfiltered
+            OrUnknown -> go filterOr' (element:filtered) unfiltered
+
+{-| 'crossProductGeneric' makes all pairs between the elements of two ors,
+then applies the given function to the result.
+
+As an example,
+
+@
+crossProductGeneric
+    f
+    (make [a1, a2])
+    (make [b1, b2])
+    ]
+@
+
+will produce something equivalent to
+
+@
+makeGeneric
+    [ f(a1, b1)
+    , f(a1, b2)
+    , f(a2, b1)
+    , f(a2, b2)
+    ]
+@
+
+-}
+crossProductGeneric
+    :: (child1 -> child2 -> child3)
+    -> MultiOr child1
+    -> MultiOr child2
+    -> MultiOr child3
+crossProductGeneric joiner (MultiOr first) (MultiOr second) =
+    MultiOr $ map (uncurry joiner) $ List.crossProduct first second
```

Comment:
This is often written one of these ways:
```.hs
map (uncurry joiner) $ List.crossProduct first second
===
-- Sometimes you see `liftA2` instead of `liftM2`; they do the same thing, 
-- but `liftA2` has a relaxed class constraint.
liftM2 joiner first second
===
-- This is probably the most common variant.
joiner <$> first <*> second
```
The second and third variants do less allocation at runtime.